code/game/machinery/pipe/construction.dm

```diff
@@ -137,23 +150,43 @@ Buildable meters
 
 	var/obj/machinery/atmospherics/fakeA = pipe_type
 	var/flags = initial(fakeA.pipe_flags)
-	for(var/obj/machinery/atmospherics/M in loc)
-		if((M.pipe_flags & flags & PIPING_ONE_PER_TURF)) //Only one dense/requires density object per tile, eg connectors/cryo/heater/coolers.
+	var/pipe_count = 0
+	for(var/obj/machinery/atmospherics/machine in loc)
+		if(machine.piping_layer != piping_layer)
+			continue
+		pipe_count += 1
+	for(var/obj/machinery/atmospherics/machine in loc)
+		if((machine.pipe_flags & flags & PIPING_ONE_PER_TURF)) //Only one dense/requires density object per tile, eg connectors/cryo/heater/coolers.
 			to_chat(user, "<span class='warning'>Something is hogging the tile!</span>")
 			return TRUE
-		if((M.piping_layer != piping_layer) && !((M.pipe_flags | flags) & PIPING_ALL_LAYER)) //don't continue if either pipe goes across all layers
+
+		if(pipe_count == 1 && istype(machine, /obj/machinery/atmospherics/pipe/smart) && pipe_type == /obj/machinery/atmospherics/pipe/smart && lowertext(machine.pipe_color) != lowertext(pipe_color) && machine.connection_num < 3)
```

Comment:
Why are you doing a direct type comparison here? You really should not be doing this sort of thing.

Reply:
Changed to ispath(pipe_type, path)

Unless I misunderstood something